### ISSUE TITLE
codedoc serve doesn't work

### ISSUE THREAD

**Issue by mhamri** [CONNECT-platform/codedoc#97]
I followed the documentation and unfortunately, it doesn't work 

![image](https://user-images.githubusercontent.com/8991783/106094579-3432b500-616d-11eb-9015-e79d586f7d05.png)

I believe the picture is quite explanatory as I did it in order to be readable. 

let me know if you need any information

**Comment by loreanvictor** [comment]
thanks for the report. may I ask which terminal software are you using? could you try on the default command prompt and see if the issue persists?

the issue is that dependencies of the CLI (in this case, `ts-node-dev`) are not by default added to `PATH`. codedoc CLI is supposed to add it for the environment of the corresponding child process, however in some environments it does not work. a quick remedy is to install `ts-node` and `ts-node-dev` explicitly (I suspect I should add that to the docs). and I suspect this is the same issue as [this](https://github.com/CONNECT-platform/codedoc/issues/85).

**Comment by mhamri** [comment]
it's the windows terminal with Powershell core 

this is the cmd that result in the same thing
![image](https://user-images.githubusercontent.com/8991783/106223729-f1301a80-621c-11eb-978a-a1a32f66ab34.png)


but installing only the `ts-node-dev` (no need `ts-node`) solved the issue, maybe worthy to investigate it more, or update the documentation 

![image](https://user-images.githubusercontent.com/8991783/106223896-453aff00-621d-11eb-8f23-fd6f525cfddc.png)

**Comment by loreanvictor** [comment]
yes, it definitely is at it has recurred multiple times. I would particularly appreciate if you were up for investigating this further, since I cannot reproduce it with Vagrant or on windows machines available to me (which are pretty limited during the quarantine, unfortunately).

this is the issue:
- the CLI [uses `ts-node-dev`](https://github.com/CONNECT-platform/codedoc-cli/blob/45ef4fdb14d849e72a431e2e8ac1add70dd7a774/src/serve.js#L22) ([and `ts-node`](https://github.com/CONNECT-platform/codedoc-cli/blob/45ef4fdb14d849e72a431e2e8ac1add70dd7a774/src/serve.js#L20))
- as a result, the CLI [installs them](https://github.com/CONNECT-platform/codedoc-cli/blob/45ef4fdb14d849e72a431e2e8ac1add70dd7a774/package.json#L33-L34) when it is installed.
- when NPM installs some package globally, it will add [its executable](https://github.com/CONNECT-platform/codedoc-cli/blob/45ef4fdb14d849e72a431e2e8ac1add70dd7a774/package.json#L6-L8) to `PATH`
- NPM does NOT add the executables of dependencies to `PATH`.
- so we need to [add the executables of dependencies (`<where-cli-is-installed>/node_modules/.bin/`) to `PATH` for child processes executed by the CLI](https://github.com/CONNECT-platform/codedoc-cli/blob/45ef4fdb14d849e72a431e2e8ac1add70dd7a774/src/exec.js#L6-L14).
- sometimes, and for some reason, this doesn't work (for example, in your case).

so far, all reports on this not working where on windows machines. even more curious, not all windows machines suffer from this issue (e.g. I am unable to replicate it).

in the meanwhile, I will add a hint to the docs to install `ts-node-dev` until this issue is resolved.

**Comment by mhamri** [comment]
@loreanvictor 
I noticed two things, in all the other issues reported in here, all of us including me, are using **nvm**

second, you are calling the `ts-node` and `ts-node-dev` as a stand-alone commands. 

![image](https://user-images.githubusercontent.com/8991783/106575688-51092700-6577-11eb-8d14-fc6ea0246988.png)

this means that this command should be registered under `PATH` which isn't. 

I can think of two solutions. 

1- this is the painful one, just require the lib and instantiate it by yourself 
```
const tsNodeDev = require("ts-node-dev");
```

this requires to look into the library implementation which is hard to maintain

2- let the `npx` execute the file

npx is a global command and part of npm. so it's safe to assume it's always there. npx look into local first, if can't' find it, it will look into the global and even if can't find it yet, it will download it and execute it. so either way that package would get run. I'm not sure of the impact but it worked when I tried it.

```
await exec(
        "npx ts-node",
        "-T --project .codedoc/tsconfig.json .codedoc/watch"
      );
```

**Comment by loreanvictor** [comment]
> this means that this command should be registered under PATH which isn't.

[It actually is](https://github.com/CONNECT-platform/codedoc-cli/blob/45ef4fdb14d849e72a431e2e8ac1add70dd7a774/src/exec.js#L6-L14), which is how it works in other operating systems (and in windows environments I try as well). For whatever reason, this does not take effect sometimes.

> just require the lib and instantiate it by yourself

That was my first goto, and I don't recall exactly why I ended up invoking the command, but I suspect ts-node-dev was not really happy with being called programmatically or something.

>  let the `npx` execute the file

This is a nice idea with one caveat: since `npx` prioritizes local version, it might run a wrong version of `ts-node` / `ts-node-dev`. The correct version will always be those installed locally on the CLI install path.

On this front: even current implementation would prioritize a global installation of `ts-node` / `ts-node-dev` over the local versions. Perhaps specifying the absolute binary paths would resolve this issue. I will look into that option in the next free time that I get.

**Comment by mhamri** [comment]
Oh sorry didn't notice you have a wrapper for exec, assumed it's nodejs's `exec`

turned out the env doesn't have the `PATH` key and it is `Path` 

the dication of PATH depends on the windows version. in the new version that key doesnt' exist anymore 

![image](https://user-images.githubusercontent.com/8991783/106597008-c1707200-6590-11eb-8053-1df506dc8eaa.png)

and this solved the issue.

```js
if (process.platform === "win32") {
      const nodePath = path.join(__dirname, "../node_modules");
      const binPath = path.join(nodePath, ".bin");
      const pathReplace = [process.env.PATH ? process.env.PATH : "", nodePath, binPath].filter((x) => !!x).join(";");

      const env = {
        ...process.env,
        Path: pathReplace,
        PATH: pathReplace,
      };

      /*  or do this:  */

      // if ("Path" in env) {
      //   env["Path"] = pathReplace;
      // } else {
      //   env["PATH"] = pathReplace;
      // }

      try {
        const child = exec(cmd + " " + params, { env });
        child.stdout.pipe(process.stdout);
        child.stderr.pipe(process.stderr);
        child.on("close", (status) => {
          if (status !== 0) reject();
          else resolve();
        });

        return child;
      } catch (error) {
        var err = new Error(error);
        reject(err);
      }
    }
```

also not sure why you need to add a key for `NODE_MODULE`. so I removed that one also as the node_module is already part of the path, so you don't need to reference to it anymore

**Comment by loreanvictor** [comment]
> also not sure why you need to add a key for `NODE_MODULE`.

Honestly I don't recall precisely, I just remember that something would go wrong somewhere in some cases 😅😅. I suspect some script from within the `.codedoc` module sometimes wants to import some stuff that are not available inside its own `node_modules`, or something in those lines.

Anyways thanks a lot! Could you make a PR to `codedoc-cli`? It would be lovely if you could keep the `NODE_PATH` for now until further investigation.

**Comment by mhamri** [comment]
sure, which of these two variations do you prefer? 

```
 const env = {
        ...process.env,
        Path: pathReplace,
        PATH: pathReplace,
      };

      /*  or do this:  */

      // if ("Path" in env) {
      //   env["Path"] = pathReplace;
      // } else {
      //   env["PATH"] = pathReplace;
      // }
```

**Comment by loreanvictor** [comment]
I suspect the first option is the safer one, as some other tooling / integration / etc might also take a peek at `PATH` variable.

**Comment by TysonMN** [comment]
Thanks @mhamri for all your hard work!  I am very happy that this bug is being fixed :D

**Comment by mhamri** [comment]
@loreanvictor may I know what formatter are you using? I'm using prettier but it changes things up. or if it's prettier, would you share your config with me?

**Comment by loreanvictor** [comment]
@mhamri nothing at the moment for codedoc (the project was from before I starting warming up to linters). generally I use eslint with `eslint:recommended` (with perhaps some additional rules or modifications here and there).

**Comment by mhamri** [comment]
done, let me know if you want me to apply any changes

**Comment by loreanvictor** [comment]
perfect. I have wrote some comments (minor stuff mostly). thanks for finally tracking this down (I would never imagined that different windows versions would use different keys for `PATH` environment variable, which is a standard on almost all other operating systems. shady move by Microsoft 😂).

**Comment by mhamri** [comment]
I think they are suffering a lot from that 

![image](https://user-images.githubusercontent.com/8991783/106635272-9bac9280-65bb-11eb-983e-6277053b77eb.png)

**Comment by loreanvictor** [comment]
so thanks to @mhamri , this issue should be fixed on `@codedoc/cli@0.2.8`. please reopen this issue if it is not.
